### PR TITLE
Add RandomForest classifier for crystals

### PR DESCRIPTION
**Pull request by christofid**

### CODE REVIEW

File: src/gt4sd/frameworks/crystals_rfc/FeatureEngine.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    def make_features(
+        self,
+        atomic_descriptors,
+        formula_list,
+        targets,
+        encoded_sym,
+        addAVG: bool = True,
+        addAAD: bool = True,
+        addMD: bool = False,
+        addCV: bool = False,
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/FeatureEngine.py:43]: would be good to use snake case and not camel

File: src/gt4sd/frameworks/crystals_rfc/FeatureEngine.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    ) -> List[List[Any]]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/FeatureEngine.py:44]: not sure I get all types right, but this Any could maybe be a Union of some types?

File: src/gt4sd/frameworks/crystals_rfc/rfclassifier.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from sklearn.impute import SimpleImputer
+from sklearn.metrics import classification_report
+from sklearn.model_selection import cross_val_score, train_test_split
+
+from .FeatureEngine import Features
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+seed = 7
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/rfclassifier.py:43]: risky :), maybe we could let it be configurable via env variables

File: src/gt4sd/frameworks/crystals_rfc/rfclassifier.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+from typing import Any, List, Tuple
+
+import numpy as np
+import pandas as pd
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.impute import SimpleImputer
+from sklearn.metrics import classification_report
+from sklearn.model_selection import cross_val_score, train_test_split
+
+from .FeatureEngine import Features
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/rfclassifier.py:38]: I would rename the module from `FeatureEngine` to `feature_engine`

File: src/gt4sd/frameworks/crystals_rfc/rfclassifier.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            df = df.loc[df[7] == 1]
+        elif self.crystal_sys == "all":
+            # Selecting  the all materials
+            pass
+
+        return df
+
+    def split_data(
+        self, df: pd.DataFrame, test_size: float = 0.2
+    ) -> Tuple[Any, Any, Any, Any]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/rfclassifier.py:113]: `Any` to `pd.DataFrame` or `np.ndarray`?

File: src/gt4sd/frameworks/crystals_rfc/rfclassifier.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        """
+        data_X, data_y = df.iloc[:, 1:].values, df.iloc[:, 0].values
+
+        train_x, test_x, train_y, test_y = train_test_split(
+            data_X, data_y, test_size=test_size, random_state=42
+        )
+
+        return train_x, test_x, train_y, test_y
+
+    def normalize_data(self, train_x:Any, test_x:Any, train_y:Any, test_y:Any) -> Tuple[Any, Any, Any, Any]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/rfclassifier.py:131]: `Any` to `pd.DataFrame` or `np.ndarray`?

File: src/gt4sd/frameworks/crystals_rfc/rfclassifier.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        )
+
+        scores = cross_val_score(clf, x, y, cv=10, scoring="accuracy")
+        model = clf.fit(x, y)
+
+        logger.info("Mean Accuracy: %0.5f (+/- %0.5f)" % (scores.mean(), scores.std()))
+
+        return model
+
+    def save(self, path: str, model: RandomForestClassifier, maxm) -> None:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/crystals_rfc/rfclassifier.py:179]: max should be `np.ndarray` or something iterable right?

**Review by drugilsberg — COMMENTED** [review]
An initial review. amazing job up to this point.

File: src/gt4sd/properties/tests/test_properties.py [hunk 8]
```diff
@@ -130,6 +130,29 @@ def test_crystals(property_key: str):
         prediction = pred_dict["1000041"]
         assert np.isclose(prediction, crystal_ground_truths[property_key], atol=1e-2)
 
+def test_rfc():
+    property_class, parameters_class = CRYSTALS_PROPERTY_PREDICTOR_FACTORY["meta_nonmetal_classifier"]
```
Comment (jannisborn) [inline src/gt4sd/properties/tests/test_properties.py:134]: meta_nonmetal_classifier --> typo?

**Review by jannisborn — COMMENTED** [review]
Great work! What do the columns in [crf_data.csv](https://github.com/GT4SD/gt4sd-core/pull/191/files#diff-71d88993ee709023b83005bb60946f2b8f0dcff6482c56e9ffdfe30ffc209cdd) correspond to? We should make inference as easy as possible, so I'm wondering whether the input could also just be a string or a list of strings (as in the first column of that file)


There are some strange issues with the build that we have never seen before (in the env setup we could not parse the RT requirement `selfies>=0.2.4<=1.0.0`. This has **always** worked since the initial release. I adapted the RT code to `selfies>=0.2.4,<=1.0.0`, but I expect further issues

File: src/gt4sd/frameworks/crystals_rfc/feature_engine.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                    num = int(num)
+                    avg += des[elem] * num
+                    des_list.append((des[elem], num))
+                avg = avg / total
+                if add_avg:
+                    feature_list.append(avg)
+
+                # Calculating Average Absolute Deviation
+                if add_aad:
+                    avgAD = 0.0
```
Comment (jannisborn) [inline src/gt4sd/frameworks/crystals_rfc/feature_engine.py:100]: no snake case

**Review by jannisborn — APPROVED** [review]
LGTM! Some typos left but not sure it's even worth re-triggering the CI

**Review by drugilsberg — CHANGES_REQUESTED** [review]
@christofid amazing job, just implement the changes @jannisborn suggested and we are good to go

**Review by drugilsberg — APPROVED** [review]

**Review by drugilsberg — CHANGES_REQUESTED** [review]
@christofid last change, can we rename the module name in `rf_classifier`?

image

**Review by drugilsberg — APPROVED** [review]
Looks, good when the tests are passing I will merge.

**Pull request merged**